explain-subqueries.md

```diff
@@ -0,0 +1,146 @@
+
+In this example, TiDB choses to use a different strategy:
+
+```sql
+EXPLAIN SELECT * FROM t1 WHERE id IN (SELECT t1_id FROM t2 WHERE t1_id != t1.int_col);
+```
+
+```sql
++-----------------------------+-----------+-----------+------------------------------+--------------------------------------------------------------------------------------------------------+
+| id                          | estRows   | task      | access object                | operator info                                                                                          |
++-----------------------------+-----------+-----------+------------------------------+--------------------------------------------------------------------------------------------------------+
+| MergeJoin_9                 | 45446.40  | root      |                              | semi join, left key:test.t1.id, right key:test.t2.t1_id, other cond:ne(test.t2.t1_id, test.t1.int_col) |
+| ├─IndexReader_24(Build)     | 180000.00 | root      |                              | index:IndexFullScan_23                                                                                 |
+| │ └─IndexFullScan_23        | 180000.00 | cop[tikv] | table:t2, index:t1_id(t1_id) | keep order:true                                                                                        |
+| └─TableReader_22(Probe)     | 56808.00  | root      |                              | data:Selection_21                                                                                      |
+|   └─Selection_21            | 56808.00  | cop[tikv] |                              | ne(test.t1.id, test.t1.int_col)                                                                        |
+|     └─TableFullScan_20      | 71010.00  | cop[tikv] | table:t1                     | keep order:true                                                                                        |
++-----------------------------+-----------+-----------+------------------------------+--------------------------------------------------------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+Here we can see that TiDB is using a `semi join`. Semi-join differs from an `INNER JOIN`, in that it will only permit the first value on the right key (`t2.t1_id`). i.e. the duplicates are eliminated as part of the join operator. The join algorithm is also a Merge Join, which is like an efficient zipper-merge as it reads data from both the left and the right side in sorted order.
+
+The query is said to be a _correlated subquery_ because the column `t1.int_col` is pushed directly into the subquery, but actually from the `EXPLAIN` output we can see that it is not correlated at all. This is because TiDB has rewritten `t1_id != t1.int_col` to `t1.id != t1.int_col` and moved it out of the subquery. This is a lot more efficient to process, because TiDB can perform this check in `└─Selection_21` as it is reading from the table `t1`.
```

Comment:
+ `is said to be` seems a bit weird.
+ We can reduce the use of `result + because`, especially in the same passage.
```suggestion
The column `t1.int_col` is pushed directly into the subquery, so this semi-join operation might seem a kind of correlated subquery. However, from the `EXPLAIN` output above, you can see that this column is not correlated at all, because TiDB has rewritten `t1_id != t1.int_col` to `t1.id != t1.int_col` and moved it out of the subquery. Then, TiDB can perform this check in `└─Selection_21` as it is reading data from the table `t1`, which makes the execution a lot more efficient.
```

Reply:
I will edit it to be clearer. It starts out correlated, but TiDB applies decorrelation as a logical optimization.

Reply:
I have updated the wording and included a link to the optimization (subquery decorrelation). PTAL again, thx :-)